registry/storage/manifeststore.go

```diff
@@ -13,6 +15,149 @@ import (
+	manifestRevisionEnumOrphaned
+	manifestRevisionEnumAll
+)
+
+type byManifestName []distribution.Manifest
+
+func (by byManifestName) Len() int      { return len(by) }
+func (by byManifestName) Swap(i, j int) { by[i], by[j] = by[j], by[i] }
+func (by byManifestName) Less(i, j int) bool {
+	return cmpManifests(by[i], by[j]) < 0
+}
+
+// cmpManifests compares two manifests and returns:
+//  -1 if the 1st shall precede the 2nd
+//   0 if manifests are the same
+//   1 if the 1st shall come after the 2nd
+// The 1st will precede the 2nd if:
+//   1. it has older schema version
+//   2. its name lexicographically precedes the other      [schema1]
+//   3. its tag lexicographically precedes the other       [schema1]
+//   4. its mediatype lexicographically precedes the other [schema2]
+//   5. its first descriptor precedes the other            [manifestlist]
+//   6. its digest lexicographically precedes the other
+func cmpManifests(fst, snd distribution.Manifest) int {
+	var (
+		fstv, sndv     manifest.Versioned
+		fstv1, sndv1   *schema1.Manifest
+		fstv2, sndv2   *schema2.Manifest
+		fstv2l, sndv2l *manifestlist.ManifestList
+	)
+
+	switch t := fst.(type) {
+	case *schema1.SignedManifest:
+		fstv1 = &t.Manifest
+		fstv = fstv1.Versioned
+	case *schema2.DeserializedManifest:
+		fstv2 = &t.Manifest
+		fstv = fstv2.Versioned
+	case *manifestlist.DeserializedManifestList:
+		fstv2l = &t.ManifestList
+		fstv = fstv2l.Versioned
+	}
+
+	switch t := snd.(type) {
+	case *schema1.SignedManifest:
+		sndv1 = &t.Manifest
+		sndv = sndv1.Versioned
+	case *schema2.DeserializedManifest:
+		sndv2 = &t.Manifest
+		sndv = sndv2.Versioned
+	case *manifestlist.DeserializedManifestList:
+		sndv2l = &t.ManifestList
+		sndv = sndv2l.Versioned
+	}
+
+	if fstv.SchemaVersion < sndv.SchemaVersion {
+		return -1
+	}
+	if fstv.SchemaVersion > sndv.SchemaVersion {
+		return 1
+	}
+	if fstv.MediaType < sndv.MediaType {
+		return -1
+	}
+	if fstv.MediaType > sndv.MediaType {
+		return 1
+	}
+
+	// manifests of schema 1
+	if fstv.SchemaVersion == 1 {
+		if fstv1.Name < sndv1.Name {
+			return -1
+		}
+		if fstv1.Name > sndv1.Name {
+			return 1
+		}
+		if fstv1.Tag < sndv1.Tag {
+			return -1
+		}
+		if fstv1.Tag > sndv1.Tag {
+			return 1
+		}
+
+		fstdgst := digest.FromBytes(fst.(*schema1.SignedManifest).Canonical)
+		snddgst := digest.FromBytes(snd.(*schema1.SignedManifest).Canonical)
+		if fstdgst.String() < snddgst.String() {
+			return -1
+		}
+		if fstdgst.String() > snddgst.String() {
+			return 1
+		}
+		return 0
+	}
+
+	// manifest lists
+	if fstv.MediaType == manifestlist.MediaTypeManifestList {
+		if len(fstv2l.Manifests) == 0 && len(sndv2l.Manifests) > 0 {
+			return -1
+		}
+		if len(fstv2l.Manifests) > 0 && len(sndv2l.Manifests) == 0 {
+			return 1
+		}
+		return cmpDescriptors(fstv2l.Manifests[0].Descriptor, sndv2l.Manifests[0].Descriptor)
+	}
+
+	// manifests of schema 2
+	return cmpDescriptors(fstv2.Config, sndv2.Config)
```

Comment:
The comparison functions are needed to sort manifest in `Enumerate` method because of parameter `last`. In the end, we need to compare revision digests here which was refused for blob enumeration as it makes no sense to return them lexicographically sorted.

This indicates flawed interface to me. If we're going to stick with returning `distribution.Manifest`, I'd go for callback interface instead:

    Enumerate(ctx context.Context, ingester func(distribution.Manifest) error) error